cpp/src/fil/fil.cu

```diff
@@ -625,13 +604,108 @@ int max_depth(const tl::ModelImpl<T, L>& model)
   return depth;
 }
 
-inline void adjust_threshold(
+// constructs a vector of size max_fid (number of features, or columns) from a Treelite tree,
+// where each feature has a maximum matching category and number of categorical nodes
+template <typename T, typename L>
+inline std::vector<int> max_matching_cat(const tl::Tree<T, L>& tree, int max_fid)
+{
+  std::vector<int> res(max_fid);
+  std::stack<int> stack;
+  stack.push(tree_root(tree));
+  while (!stack.empty()) {
+    int node_id = stack.top();
+    stack.pop();
+    while (!tree.IsLeaf(node_id)) {
+      if (tree.SplitType(node_id) == tl::SplitFeatureType::kCategorical) {
+        std::vector<uint32_t> mmv = tree.MatchingCategories(node_id);
```

Comment:
Don't forget to `#include <stdint.h>` or better yet `#include <cstdint>` and make this `std::uint32_t`.

Reply:
An include is necessary in order to avoid a [transitive dependency](https://godbolt.org/z/r1MoWWETs). Our current unfortunate situation with a tangled include graph and the correspondingly long parse time during compilation is a result of not adhering to "include what you use" practices. It's up to you whether you use `uint32_t` or `std::uint32_t` (though I would recommend the latter), but we need the include.

Reply:
My understanding is that the preprocessor converts the include statements to verbatim include the code, which means, that including transitive dependencies explicitly only grows the file that's handled to the parser. But let's look at the real world data.

Here are the compilation times for `fil.cu` that I've measured using this advice. As you see, it, unfortunately, including (or not including) doesn't significantly affect them. I suspect the difference and the deviation will tend to an even less significant amount.

Could you raise a proper discussion about this, so that I know there isn't another reviewer giving the exact opposite advice? Because from the other uses of `uint32_t`, almost all adhere to the convention used in this PR.

|| user time    |             |                             |   | real time    |             |                             |
|-------|--------------|-------------|-----------------------------|---|--------------|-------------|-----------------------------|
|...| with include | w/o include | preventable delay           |   | with include | w/o include | preventable delay           |
|run 1 | 11.731 s     | 11.605 s    |                             |   | 12.478 s     | 12.395 s    |                             |
|run 2 | 11.585 s     | 11.546 s    |                             |   | 12.395 s     | 12.377 s    |                             |
|run 3 | 11.658 s     | 11.592 s    |                             |   | 12.469 s     | 12.454 s    |                             |
|run 4 | 11.600 s     | 11.559 s    |                             |   | 12.397 s     | 12.352 s    |                             |
|average | 11.64 s      | 11.58 s     | −0.07 (slower with include) |   | 12.43 s      | 12.39 s     | −0.04 (slower with include) |
|stdev of avg | 0.04 s       | 0.02 s      | 0.04                        |   | 0.03 s       | 0.03 s      | 0.04                        |